src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala

```diff
@@ -35,24 +35,50 @@ abstract class Pickler extends SubComponent {
   class PicklePhase(prev: Phase) extends StdPhase(prev) {
     def apply(unit: CompilationUnit) {
       def pickle(tree: Tree) {
-        def add(sym: Symbol, pickle: Pickle) = {
+        def add(sym: Symbol, pickle: Pickle) = withPickle(sym, pickle){ pickle =>
+          debuglog("pickling " + sym)
+          pickle putSymbol sym
+          currentRun.symData(sym) = pickle
+        }
+
+        def withPickle(sym: Symbol, pickle: Pickle)(f: Pickle => Unit) =
           if (currentRun.compiles(sym) && !currentRun.symData.contains(sym)) {
-            debuglog("pickling " + sym)
-            pickle putSymbol sym
-            currentRun.symData(sym) = pickle
+            f(pickle)
           }
-        }
 
         tree match {
           case PackageDef(_, stats) =>
             stats foreach pickle
           case ClassDef(_, _, _, _) | ModuleDef(_, _, _) =>
             val sym = tree.symbol
-            val pickle = new Pickle(sym)
-            add(sym, pickle)
-            add(sym.companionSymbol, pickle)
-            pickle.writeArray()
-            currentRun registerPickle sym
+
+            def shouldPickle(sym: Symbol) = currentRun.compiles(sym) && !currentRun.symData.contains(sym)
+            val shouldPickleSym = shouldPickle(sym)
+            if (shouldPickleSym) {
+              val pickle = new Pickle(sym)
+              def reserveDeclEntries(sym: Symbol): Unit = {
```

Comment:
How about:
```
def reserveDeclEntries(sym: Symbol): Unit =
  if (sym.isClass) sym.info.decls.foreach(reserveDeclEntries)
  else if (sym.isModule) reserveDeclEntries(sym.moduleClass)
  else pickle.reserveEntry(sym)
```
?

Reply:
Actually, we need to reserve the slots for the class/module symbols themselves, prior to recursing into their decls. Pushed.